pkg/distributor/query.go

```diff
@@ -232,22 +255,53 @@ func (d *Distributor) queryIngesterStream(ctx context.Context, replicationSet ri
 					}
 				}
 
-				if chunkBytesLimitErr := queryLimiter.AddChunkBytes(resp.ChunksSize()); chunkBytesLimitErr != nil {
+				if chunkBytesLimitErr := queryLimiter.AddChunkBytes(ingester_client.ChunksSize(resp.Chunkseries)); chunkBytesLimitErr != nil {
 					return ingesterQueryResult{}, validation.LimitError(chunkBytesLimitErr.Error())
 				}
 
 				result.chunkseriesBatches = append(result.chunkseriesBatches, resp.Chunkseries)
+			} else if len(resp.Series) > 0 {
+				labelsBatch := make([]labels.Labels, 0, len(resp.Series))
+				streamingSeriesCount += len(resp.Series)
+
+				for _, s := range resp.Series {
+					if limitErr := queryLimiter.AddSeries(s.Labels); limitErr != nil {
+						return ingesterQueryResult{}, validation.LimitError(limitErr.Error())
+					}
+
+					labelsBatch = append(labelsBatch, mimirpb.FromLabelAdaptersToLabels(s.Labels))
+				}
+
+				streamingSeriesBatches = append(streamingSeriesBatches, labelsBatch)
+			}
+
+			if resp.IsEndOfSeriesStream {
+				if streamingSeriesCount > 0 {
+					result.streamingSeries.Series = make([]labels.Labels, 0, streamingSeriesCount)
+
+					for _, batch := range streamingSeriesBatches {
+						result.streamingSeries.Series = append(result.streamingSeries.Series, batch...)
+					}
+
+					streamReader := ingester_client.NewSeriesChunksStreamReader(stream, streamingSeriesCount, queryLimiter)
+					closeStream = false
+					result.streamingSeries.StreamReader = streamReader
+				}
```

Comment:
[nit] I would reset the streaming series batches and count too. They're not used anymore. We can release the streamingSeriesBatches memory reference (since the function hasn't done yet at this point and it will keep allocating later on) and we can reset the count too for consistency.

```suggestion
				    streamingSeriesBatches = nil
				    streamingSeriesCount = 0
				}
```

Reply:
Given we return on the next line, do we need to release `streamingSeriesBatches`? `streamingSeriesBatches` is scoped to the function that is about to return.